File: frequency_of_char_in_string/C-sharp/frequency_of_char_in_string.cs
using System;
// use this thing to make dictionary
using System.Collections.Generic;

namespace FreqCharInStr
{
	class Program
	{
		// make a class for implementating the function to remove the space character
		class StringManipulate
		{
			/// <summary>Receive a string then return a new string with no space in it</summary>
			/// <parameter>String str</paramter>
			/// <return>A string</return>
			public string rmSpaceChar(string str)
			{
				// make a new string so that it won't affect on the input string
				string new_str = str;
				// replace space characters in the string, eg: "Hello World" -> "HelloWorld"
				new_str = new_str.Replace(" ","");
				return new_str;
			}
		}

		// class for sorting the array in ascend order
		class ArraySorting
		{
			/// <summary>Receive a string, then put the letter in the alphabeltic order in the return string</summary>
			public string SortArray(string str)
			{
				// convert the string into array of char
				char[] my_str = str.ToCharArray();
				// sort the array ascendingly, according to the ASCII table
				Array.Sort(my_str);
				// make a string from the array of string
				string new_str = new string(my_str);
				return new_str;
			}
		}

		// get the frequency of character in string
		class Frequency
		{
			/// <summary>a method to get the freq of char in str</summary>
			/// <return>A dictionary holding the char and its freq</return>

			public Dictionary<char,int> Frequency_of_char(string str)
			{
				// make a dictionay to hold the char and its freq
				Dictionary<char,int> dict = new Dictionary<char,int>();
				// get the length of the string
				int len = str.Length;
				// count variable to travel through the array to get the char
				int i = 0;
				int j = 0;
				// the frequency of char, first initialize it with `0`, because everything in the first place should be started with `0`
				int count = 0;
				
				// loop through the array with `while` loop because the iteriatation of this loop is unpredictable
				while(i < len)
				{
					// every character should appear at least once in the array
					count = 1;
					// if `i` reach the last index, print out that this is the last character, it should appear once only in the string
					if(i == len - 1) {
						// add to the Dictionary a new pair of value
						dict.Add(str[i],count);
						// break the loop because it already got to the last index
						break;
					}else{
						// if `i` doesn't get to the last index, `j` will be the next index
						j = i + 1;
					}

					// compare two consecutive char, if they are the same, remove the second char, decrease the len of the array by 1
					// the same character means that the character `str[i]` has to appear more than 1, increase count by 1
					while(str[i] == str[j])
					{
						count++;
						str = str.Remove(j,1);
						len--;
						// if `i` got to the last character while we remove character at index `j`, exit the loop
						// for example, some last character of the loop is the same
						/*
						*	'www' -> i = 0, j = 1, len = 3
						*	'ww'  -> i = 0, j = 1, len = 2
						* 	'w'   -> i = 0, j = 1, len = 1 -> i == len - 1, this is the end of the loop, break
						* If not, `j` will be j = 1, index out of bound exception
						*/
						if(i == len - 1)
						{
							break;
						}
					}

					// add to the Dictionary a new pair of char and its frequency
					dict.Add(str[i],count);
					// increase `i`(index) so that we won't be arrange in infinite loop
					i++;
				}

				// bring all the dictionary back to the main function
				return dict;
			}
		}

		static void Main(string[] args)
		{
			// make a string
			string input = "Hello C sharp world";
			
			/// <heading>Remove space character</heading>
			// make an object to implement the function
			StringManipulate rmSpaceChar = new StringManipulate();
			// make an empty string to receive a string without spaces
			string my_str = "";
			// remove the space then assign it
			my_str = rmSpaceChar.rmSpaceChar(input);

			/// <heading>Convert the all uppercase letter in the string into lowercase</heading>
			my_str = my_str.ToLower();
			
			/// <heading>Sort the string</heading>
			ArraySorting arrSort = new ArraySorting();
			// sort the array
			my_str = arrSort.SortArray(my_str);

			/// <heading>Calculate the frequency of character in the string</heading>
			// create an instance to implement the method to count the frequency
			Frequency freq = new Frequency();
			// make a dictionary
			Dictionary<char,int> my_dict = new Dictionary<char,int>();
			// initialize to the dictionary
			my_dict = freq.Frequency_of_char(my_str);

			// print out the dictionary, each key-value
			// make a KeyValuePair instance to access to the pair of values in Dictionary
			foreach(KeyValuePair<char,int> word in my_dict)
			{
				// Access the `Key` and `Value` field of the `word` instance
				Console.WriteLine("\'{0}\' : {1}",word.Key,word.Value);
			}
		}
	}
}

File: frequency_of_char_in_string/VueJS/frequency_of_char_in_string.js
function stringFormat(the_string)
{
	// <heading> Remove all the space character with nothing</heading>
	// to remove all character, we have to use regular expression
	// For example, / /g -> all spaces, /blue/g -> all the string "blue"
	var no_space_str = the_string.replace(/ /g,"");

	// <heading> Convert the string to the lowercase string with `toLowerCase` </heading>
	var lwr_str = no_space_str.toLowerCase();

	// <heading> Convert the string into array with `split`, distinguish different element with nothing("") </heading>
	// This is just like Perl
	var arr = lwr_str.split("");

	// <heading> Sort the array </heading>
	// this is why we have to convert it to the array, we cannot sort the string
	var sort_arr = arr.sort();

	// // <heading> Convert the array of char back to the string </heading>
	var sort_str = sort_arr.join("");

	return sort_str;
}

// make an Object, receive two parameter, a character(`char`) and its frequency(`freq`)
function Dictionary(char,freq)
{
	// assign the `char` and `freq` from the paramter to them of the Object (this.char = char)
	this.char = char;
	this.freq = freq;
}

// print the dict to the console, push it into the array `dictionary` in the `app`(VueJS instance)
// so that VueJS can log it out the web page
function printDict(dict)
{
	// get the char and freq of the Dictionary, then assign to local variable
	var char = dict.char;
	var freq = dict.freq;
	// pass the data to the console, which is actually, push new record into the dictionary array
	// `char` and `freq` will automatically be replaced with actual value
	// This is one of the most fantastic thing in VueJS, the console pipe
	console.log(app.dictionary.push({character:char,frequency:freq}));
}

// calculate the frequency of character in the string
function freqOfChar(my_str)
{
	// convert the array to the string, with `split`, every char of string will an element(`split("")`)
	var arr = my_str.split("");
	// get the number of element in the array with `length`
	var len = arr.length;
	// make an empty array, with the size is like array `arr`
	var dict = new Array(len);

	// some run variable for the while-loop
	var i = 0;
	var j = 0;
	// the original frequency of the char
	var count = 0;

	// loop through each element of the `arr`
	while(i < len)
	{
		// each character appears at least one in the string, so that the default value must be 1
		count = 1;
		// if we reach the last character of the string, this char will appear only once in the string
		// so that make a record of character (`arr[i]`) and default freq(count = 1)
		if(i == len - 1) {
			dict[i] = new Dictionary(arr[i],count);
			// exit the loop, prevent the below code(which is only applied when we haven't reached the last element) from happening
			break;
		}else{
			// if we haven't reached the last element, get the next-to character
			j = i + 1;
		}

		// while two next-to characters are the same
		// increase the freq(`count++`), remove the second element(which is at index `j`) (`arr.splice(j,1)`)
		// decrease the len(`len--`) and if while the `len` is reduced, it make `i` reach the last element, exit the current while, print out the freq
		while(arr[i] == arr[j])
		{
			count++;
			arr.splice(j,1);
			len--;
			if(i == len - 1)
			{
				break;
			}
		}

		// assign to the array of Dictionary, a new record of the current char and its frequency
		dict[i] = new Dictionary(arr[i],count);
		// increase the run variable `i` so that we won't catch the infinite loop
		i++;
	}

	// give all the array of Dictionary back to the caller(main function)
	return dict;
}

// the ultimate function, receive the string, and print each character and its frequency to the console
// so that VueJS can take that and present to the web page
function calcFreqOfChar(the_str)
{
	// formate the given string(remove spaces, uppercase to lowercase, sort)
	var new_str = stringFormat(the_str);
	// `split` each character in the string into individual element of the array
	var arr = new_str.split("");
	// get the number of elements in the array
	var len = arr.length;
	// make a empty array that will hold the array of Dictionary(instance to store character and its frequency)
	var new_dict = new Array(len);
	// assign to it
	new_dict = freqOfChar(new_str);
	// `forEach` element inside the Dictionary array, print out the element to the console, with some formats to push into the `app.dictionary` array
	new_dict.forEach(printDict);
}

// make a new HTML Component, for looping through the array(app.dictionary) and print out the char and freq with format : 'char':freq
Vue.component('paragraph', {
	// v-bind:record -> receive a record instance
	props:['record'],
	// access the record instance in the field of `character` and `frequency`
	template: '<p>\'{{ record.character }}\' : {{ record.frequency }}'
})

// declare a Vue instance, applied of element whose id = app
var app = new Vue({
	el: "#app",
	// the data, infomation in the `app`, replace the value in `{{value}}`
	data: {
		// reflect will be in the input field and the `p{{reflect}}`
		reflect: "Hello World",
		// an empty array, waiting for being push data
		dictionary: [
		]
	},
	// an method that is in the `app`
	methods: {
		// a function called when the user type anything in the input
		freqOfCharStr: function() {
			// remove all the element in the `dictionary` array
			this.dictionary.splice(0,this.dictionary.length);
			// get the text of the input
			var text = this.reflect;
			// calculate the frequency of the given string, then log out to the web page
			calcFreqOfChar(text);
		}
	}
})
